core/src/main/java/io/micronaut/core/io/service/MicronautMetaServiceLoaderUtils.java

```diff
@@ -0,0 +1,338 @@
+        }
+        List<URI> resourceDefs = IOUtils.getResources(classLoader, MICRONAUT_SERVICES_PATH);
+        if (resourceDefs.isEmpty()) {
+            return Map.of();
+        }
+
+        Map<String, Set<String>> services = new LinkedHashMap<>();
+
+        FileVisitor<Path> visitor = new FileVisitor<>() {
+
+            private Set<String> definitions;
+
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                if (dir.endsWith(MICRONAUT_SERVICES_PATH)) {
+                    return FileVisitResult.CONTINUE;
+                }
+                String serviceName = dir.getFileName().toString();
+                definitions = services.get(serviceName);
+                if (definitions == null) {
+                    definitions = new LinkedHashSet<>();
+                    services.put(serviceName, definitions);
+                }
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path currentPath, BasicFileAttributes attrs) throws IOException {
+                if (Files.isHidden(currentPath)) {
+                    return FileVisitResult.CONTINUE;
+                }
+                Path fileName = currentPath.getFileName();
+                if (fileName.startsWith(".")) {
+                    return FileVisitResult.CONTINUE;
+                }
+                definitions.add(fileName.toString());
+                return FileVisitResult.SKIP_SUBTREE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+                return FileVisitResult.CONTINUE;
+            }
+        };
+
+        List<Closeable> toClose = new ArrayList<>();
+        try {
+            for (URI uri : resourceDefs) {
+                Path myPath = IOUtils.resolvePath(uri, MICRONAUT_SERVICES_PATH, toClose);
+                if (myPath != null) {
+                    Files.walkFileTree(myPath, Collections.emptySet(), 2, visitor);
+                }
+            }
+        } catch (IOException e) {
+            // ignore, can't do anything here and can't log because class used in compiler
+        } finally {
+            for (Closeable closeable : toClose) {
+                try {
+                    closeable.close();
+                } catch (IOException ignored) {
+                }
+            }
+        }
+        return services;
+    }
+
+    private static <S> S instantiate(String className, ClassLoader classLoader) {
+        try {
+            @SuppressWarnings("unchecked") final Class<S> loadedClass =
+                (Class<S>) Class.forName(className, false, classLoader);
+            // MethodHandler should more performant than the basic reflection
```

Comment:
idk about that given that there's still a lookup involved

Reply:
It's just a copy form the prev loader